File: app/src/main/java/com/example/mhdstuff/activity/DeparturesActivity.java
package com.example.mhdstuff.activity;

import android.content.Context;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import com.example.mhdstuff.R;
import com.example.mhdstuff.activity.base.BaseActivity;
import com.example.mhdstuff.activity.data.DelaysDataHolder;
import com.example.mhdstuff.parsing.storage.IdStorage;
import com.example.mhdstuff.parsing.types.Post;
import com.example.mhdstuff.parsing.types.Stop;
import com.example.mhdstuff.parsing.types.departure.Departure;
import com.example.mhdstuff.parsing.types.departure.Departures;
import com.example.mhdstuff.util.Container;
import com.example.mhdstuff.util.OfflineDepartures;
import com.google.gson.JsonObject;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;

public class DeparturesActivity extends BaseActivity {




    public DeparturesActivity() {
        super("", R.layout.activity_departures);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Stop stop = getIntent().getParcelableExtra("stop");
        setName(stop.name);

        Container<View> heartFullCont = new Container<>();
        View heartEmpty = addButtonIcon(R.drawable.heart_regular, v -> {
            v.setVisibility(View.GONE);
            heartFullCont.item.setVisibility(View.VISIBLE);

            stop.setFavourite(true);
        });

        View heartFull = addButtonIcon(R.drawable.heart_solid, v -> {
            heartEmpty.setVisibility(View.VISIBLE);
            v.setVisibility(View.GONE);

            stop.setFavourite(false);
        }, false);

        heartFullCont.item = heartFull;

        if (stop.isFavourite()) {
            heartEmpty.setVisibility(View.GONE);
        } else {
            heartFull.setVisibility(View.GONE);
        }

        Context context = this;
        JsonObject delays = DelaysDataHolder.getDelays();

        new Thread(() -> {
            IdStorage storage = IdStorage.getInstance();

            Departures departures = getOffline(storage, delays, stop.id);

            LinearLayout layout = findViewById(R.id.departure_items);

            createEntries(departures, layout, stop.id, context, storage);
        }).start();
    }


    private Departures getOffline(IdStorage storage, JsonObject delays, int stopId) {
        return new Departures(
                "Work in progress...",
                OfflineDepartures.getOffline(storage, stopId, delays)
        );
    }

    private void createEntries(Departures departures, LinearLayout layout, int stopId, Context context, IdStorage storage) {
        List<Departure> departureList = new ArrayList<>(departures.departures());
        AtomicInteger index = new AtomicInteger(1);

        // incremental loading of elements to increase speed of opening the screen; the effect is practically unnoticeable
        final int viewsPerFrame = 2;

        runOnUiThread(() -> {
            TextView message = findViewById(R.id.departure_message);
            if (departureList.isEmpty()) {
                message.setVisibility(TextView.VISIBLE);
                message.setText("Nebyly nalezeny žádné odjezdy...");

                ((GradientDrawable)message.getBackground()).setColor(ContextCompat.getColor(this, R.color.widget_background));
            } else if (!departures.message().isBlank()) {
                message.setVisibility(TextView.VISIBLE);
                message.setText(departures.message());
            }
        });

        while (!departureList.isEmpty()) {
            CountDownLatch latch = new CountDownLatch(1);

            runOnUiThread(() -> {
                for (int i = 0; i < viewsPerFrame && !departureList.isEmpty(); i++) {
                    Departure departure = departureList.remove(0);

                    View departureView = departure.createDepartureView(this, layout, context);
                    Post post = storage.postStorage().getPost(stopId, departure.postID());
                    departureView.setOnClickListener(v -> startActivity(
                            DeparturePostDetailActivity.class,
                            intent -> intent.putExtra("post", post)
                    ));

                    layout.addView(departureView, index.getAndIncrement());
                }
                latch.countDown();
            });
            try {
                latch.await();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }

    @Override
    public void onPause() {
        super.onPause();
    }
}

File: app/src/main/java/me/miran/mhdstuff/activity/DeparturesActivity.kt
package me.miran.mhdstuff.activity

import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewmodel.compose.viewModel
import com.google.gson.JsonObject
import me.miran.mhdstuff.R
import me.miran.mhdstuff.activity.base.KBaseActivity
import me.miran.mhdstuff.activity.data.DelaysDataHolder
import me.miran.mhdstuff.exception.RequestException
import me.miran.mhdstuff.parsing.storage.IdStorage
import me.miran.mhdstuff.parsing.types.departure.Departures
import me.miran.mhdstuff.parsing.types.stop.Stop
import me.miran.mhdstuff.util.OfflineDepartures
import me.miran.mhdstuff.util.Text
import me.miran.mhdstuff.util.request.RequestHelper


class DeparturesActivity : KBaseActivity("") {

    class StopViewModel : ViewModel() {
        private val _liked = mutableStateOf(false)
        private val _isRefreshing = mutableStateOf(false)
        val liked = _liked
        val refreshing = _isRefreshing

        fun toggleLiked() {
            _liked.value = !_liked.value
        }

        fun setLiked(value: Boolean) {
            _liked.value = value
        }

        fun setRefreshing(value: Boolean) {
            _isRefreshing.value = value
        }

    }

    @Composable
    override fun CreateElements() {
        val stop = intent.getParcelableExtra<Stop>("stop")!!
        val vm: StopViewModel = viewModel()

        LaunchedEffect(Unit) {
            if (stop.isFavourite) {
                vm.setLiked(true)
            }
        }

        createDepartures(stop)
    }

    fun createDepartures(stop: Stop, refreshDelays: Boolean = false ,onFinish: () -> Unit = {}) {
        var delays = DelaysDataHolder.getDelays()
        Thread {
            if (refreshDelays) {
                try {
                    delays = RequestHelper.getRouteDelays(this)
                } catch (e: RequestException) {
                    // TODO show error
//                e.showError(this, AppException.NotificationType.SNACK_BAR)
                    delays = JsonObject()
                }
            }

            val storage = IdStorage.getInstance()
            val departures = Departures(
                "Work in progress...",
                OfflineDepartures.getOffline(storage, stop.id.internal, delays)
            )

            runOnUiThread {
                setBaseContent {
                    this.Departures(departures, storage)
                }
                onFinish()
            }
        }.start()
    }

    override fun setBaseContent(
        actions: @Composable (RowScope.() -> Unit),
        content: @Composable (() -> Unit)
    ) {
        val stop = intent.getParcelableExtra<Stop>("stop")!!

        name = Text.literal(stop.name)

        super.setBaseContent({
            actions()
            val vm: StopViewModel = viewModel()
            val liked by vm.liked

            IconButton(onClick = {
                vm.toggleLiked()
                stop.setFavourite(liked)
                // FIXME flushing everytime is not ideal
                stop.flush()
            }) {
                if (liked) {
                    Icon(
                        painter = painterResource(R.drawable.heart_solid),
                        contentDescription = "Unlike",
                        tint = Color.Red,
                        modifier = Modifier.size(32.dp)
                    )
                } else {
                    Icon(
                        painter = painterResource(R.drawable.heart_regular),
                        contentDescription = "Like",
                        tint = colorResource(R.color.light_blue),
                        modifier = Modifier.size(32.dp)
                    )
                }
            }
        }, content)
    }

    @Composable
    fun Departures(departures: Departures, storage: IdStorage) {
        val stop = intent.getParcelableExtra<Stop>("stop")!!

        val vm: StopViewModel = viewModel()
        val refreshing by vm.refreshing

        PullToRefreshBox(refreshing, {
            vm.setRefreshing(true)

            createDepartures(stop, true) {
                vm.setRefreshing(false)
            }
        }) {
            LazyColumn {
                items(departures.departures) { entry ->
                    val post = storage.postStorage.getPost(stop.id.internal, entry.postID);

                    Departure(entry, post)
                }
            }
        }
    }

}